tower-http/src/set_response_header.rs

```diff
@@ -0,0 +1,457 @@
+//! # let render_html = tower::service_fn(|request: Request<Body>| async move {
+//! #     Ok::<_, std::convert::Infallible>(Response::new(Body::from("1234567890")))
+//! # });
+//! #
+//! let mut svc = ServiceBuilder::new()
+//!     .layer(
+//!         // Layer that sets `Content-Length` if the body has a known size.
+//!         // Bodies with streaming responses wont have a known size.
+//!         SetResponseHeaderLayer::new(
+//!             http::header::CONTENT_LENGTH,
+//!             |response: &Response<Body>| {
+//!                 if let Some(size) = response.body().size_hint().exact() {
+//!                     // If the response body has a known size, returning `Some` will
+//!                     // set the `Content-Length` header to that value.
+//!                     Some(HeaderValue::from_str(&size.to_string()).unwrap())
+//!                 } else {
+//!                     // If the response body doesn't have a known size, return `None`
+//!                     // to skip setting the header on this response.
+//!                     None
+//!                 }
+//!             }
+//!         )
+//!     )
+//!     .service(render_html);
+//!
+//! let request = Request::new(Body::empty());
+//!
+//! let response = svc.ready_and().await?.call(request).await?;
+//!
+//! assert_eq!(response.headers()["content-length"], "10");
+//! #
+//! # Ok(())
+//! # }
+//! ```
+
+use futures_util::ready;
+use http::{header::HeaderName, HeaderValue, Response};
+use pin_project::pin_project;
+use std::{
+    fmt,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use std::{future::Future, marker::PhantomData};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`SetResponseHeader`] which adds a response header.
+///
+/// See [`SetResponseHeader`] for more details.
+pub struct SetResponseHeaderLayer<M, T> {
+    header_name: HeaderName,
+    make: M,
+    mode: InsertHeaderMode,
+    _marker: PhantomData<fn() -> T>,
+}
+
+impl<M, T> fmt::Debug for SetResponseHeaderLayer<M, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SetResponseHeaderLayer")
+            .field("header_name", &self.header_name)
+            .field("mode", &self.mode)
+            .field("make", &std::any::type_name::<M>())
+            .finish()
+    }
+}
+
+impl<M, T> SetResponseHeaderLayer<M, T> {
+    /// Create a new [`SetResponseHeaderLayer`].
+    pub fn new(header_name: HeaderName, make: M) -> Self
+    where
+        M: MakeHeaderValue<T>,
+    {
+        Self {
+            make,
+            header_name,
+            mode: InsertHeaderMode::OverrideExisting,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Set which mode to use when inserting the header.
+    ///
+    /// Defaults to [`InsertHeaderMode::OverrideExisting`].
+    pub fn mode(mut self, mode: InsertHeaderMode) -> Self {
+        self.mode = mode;
+        self
+    }
```

Comment:
here's a thought, let me know what you think: it seems like with this API, where the layer is always created in the default mode and then a new mode is set, it's possible for users to get surprising behavior by not reading the docs carefully. we *might* want to consider an API design where the user _can't_ easily construct a new `SetResponseHeaderLayer` without being aware of the behavior of the service it produces.

One approach would be to always require the mode to be explicitly provided in `new`, which is kind of wordy and annoying --- you'd have to write out `InsertHeaderMode::OverrideExisting` every time you create a `SetResponseHeaderLayer`. Alternatively, we could replace the single `new` constructor with three constructors that build layers in each mode --- `new_overwrite`, `new_append`, and `new_if_not_present` or something (or, drop the `new_` prefix)? this is a little less wordy than taking the mode as a parameter, but it still makes it impossible to construct a layer without being aware of the behavior you're going to get.

let me know if you think this is a good idea, or just an additional speed bump that makes the API more complicated? i could be convinced either way.

Reply:
the other advantage of having three constructors is that we can make the enum private, reducing the number of types in the API surface...which is not the _biggest_ deal...

Reply:
I think its a good thing to make users more conscious about. Could be a source of bugs otherwise. And I don't think it makes the API more complicated.

I agree including the mode in `new` is a bit much. I think perhaps skipping the `new_` prefix could work. So we would have:

- `SetResponseHeader::overriding` as `override` is a keyword in Rust.
- `SetResponseHeader::appending` to match the tense of `overriding`
- `SetResponseHeader::if_not_present`. A shame this one doesn't match, but `inserting_if_not_present` is perhaps a bit much 😅 Could also be `unless_present` but I'd be fine with `if_not`. I've done a lot of ruby with its unless-statements.

Reply:
hmm, this seems reasonable...it would be nice if the tense for all three matched but i can't think of a good naming scheme where they all line up :( so this is good in my book!